### PR TITLE
Ch5471

### PR DESCRIPTION
**Pull request by lukerQuant**
- Reintegrating Bitcoin
- Bitcoin transactions can now take multiple inputs
- Bitcoin Transactions abide by the Overledger Transaction Object definition

### CODE REVIEW

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts [hunk 1]
```diff
@@ -140,7 +128,13 @@ if (!Object.values(TransactionBitcoinSubTypeOptions).includes(thisBitcoinTx.subT
   //make sure that the fee price + transaction amounts equal the input amount (minus dust??)
   //this way we can alert the user if he expected change to be given automatically!
 
-    //start here
+  if (Number(totalInputAmount)-Number(totalOutputAmount)-Number(thisBitcoinTx.extraFields.feePrice) != 0){ //providing a bit of leway for javascript parsing errors
```
[the remark this answers is not in the document]
Comment (najla-quant) [reply packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: If it is a string, might be better to use parseInt or parseFloat

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts [hunk 2]
```diff
@@ -96,13 +89,7 @@ if (!Object.values(TransactionBitcoinSubTypeOptions).includes(thisBitcoinTx.subT
...
       error: 'All transactions for Bitcoin must have the extraFields field set with feePrice parameters within it'
     } 
-  } else if ((!thisBitcoinTx.extraFields)||(thisBitcoinTx.extraFields == null)){
-    return {
-      success: false,
-      failingField: "extraFields",
-      error: 'All transactions for Bitcoin must have the extraFields field set with feePrice parameters within it'
-    } 
-  } else if ((thisBitcoinTx.extraFields.feePrice == "")||(thisBitcoinTx.extraFields.feePrice == null)||(thisBitcoinTx.extraFields.feePrice === 'undefined')||(!Number.isNaN(Number(thisBitcoinTx.extraFields.feePrice)))){
+  } else if ((thisBitcoinTx.extraFields.feePrice == "")||(thisBitcoinTx.extraFields.feePrice == null)||(thisBitcoinTx.extraFields.feePrice === 'undefined')){
```
Comment (najla-quant) [inline packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: === for all

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts [hunk 3]
```diff
@@ -50,28 +46,25 @@ class Bitcoin extends AbstractDLT {
...
-
-    const tx = new bitcoin.TransactionBuilder(this.addressType);
+    let tx;
+    tx = new bitcoin.TransactionBuilder(this.addressType, 0); //set maximum fee rate = 0 to be flexible on fee rate
     const data = Buffer.from(thisTransaction.message, 'utf8'); // Message is inserted
-
     let counter = 0;
     while (counter < thisTransaction.txInputs.length){
-      tx.addInput(thisTransaction.txInputs[counter].linkedTx, thisTransaction.txInputs[counter].linkedIndex);
+      tx.addInput(thisTransaction.txInputs[counter].linkedTx, Number(thisTransaction.txInputs[counter].linkedIndex));
```
Comment (najla-quant) [inline packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: What is the type of thisTransaction.txInputs[counter].linkedIndex ? I am afraid that Number() can convert anything into a number (even if it is not a number initially, JS is not typed and fails sometimes). Better to use parseInt or parseFloat and do a check like typeof "xxx" === "number"

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts [hunk 4]
```diff
@@ -1,15 +1,13 @@
...
 import { Options, Account, TransactionRequest, ValidationCheck} from '@quantnetwork/overledger-types';
 import TransactionBitcoinRequest from './DLTSpecificTypes/TransactionBitcoinRequest';
 import TransactionBitcoinSubTypeOptions from "./DLTSpecificTypes/associatedEnums/TransactionBitcoinSubTypeOptions";
 
-
 class Bitcoin extends AbstractDLT {
   NON_DUST_AMOUNT: number = 546;
-  addressType: string;
-  privateKey: string;
+  addressType: bitcoin.Network;
```
Comment (najla-quant) [inline packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: privateKey field is moved?

File: packages/overledger-dlt-abstract/src/AbstractDLT.ts [hunk 5]
```diff
@@ -156,7 +156,7 @@ abstract class AbstractDLT {
         } else if (thisTransaction.type == TransactionTypeOptions.utxo){
           
           let thisUtxoTx = <TransactionUtxoRequest> thisTransaction;
-          if ((!this.conversionTest(thisUtxoTx.txInputs))||(thisUtxoTx.txInputs == null)){
+          if (thisUtxoTx.txInputs == null){
```
Comment (najla-quant) [inline packages/overledger-dlt-abstract/src/AbstractDLT.ts]: ===

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts [hunk 6]
```diff
@@ -115,22 +102,23 @@ if (!Object.values(TransactionBitcoinSubTypeOptions).includes(thisBitcoinTx.subT
   let totalOutputAmount = 0;
   while (counter < thisBitcoinTx.txInputs.length){
     
-    if ((!thisBitcoinTx.txInputs[counter].amount)||(thisBitcoinTx.txInputs[counter].amount == null)){
+    if (!("amount" in thisBitcoinTx.txInputs[counter])){
```
[the remark this answers is not in the document]
Comment (najla-quant) [reply packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: Replaced by (!thisBitcoinTx.txInputs[counter].amount || thisBitcoinTx.txInputs[counter].amount === undefined) (Test related)

File: yarn.lock [hunk 7]
```diff
@@ -1782,8 +1794,7 @@ bitcoinjs-lib@4.0.5:
     merkle-lib "^2.0.10"
     pushdata-bitcoin "^1.0.1"
     randombytes "^2.0.1"
-    safe-buffer "^5.1.1"
-    tiny-secp256k1 "^1.0.0"
+    tiny-secp256k1 "^1.1.1"
```
[the remark this answers is not in the document]
Comment (najla-quant) [reply yarn.lock]: kept

File: yarn.lock [hunk 8]
```diff
@@ -8379,7 +8390,7 @@ timed-out@^4.0.0, timed-out@^4.0.1:
   resolved "https://registry.yarnpkg.com/timed-out/-/timed-out-4.0.1.tgz#f32eacac5a175bea25d7fab565ab3ed8741ef56f"
   integrity sha1-8y6srFoXW+ol1/q1Zas+2HQe9W8=
 
-tiny-secp256k1@^1.0.0:
+tiny-secp256k1@^1.1.1, tiny-secp256k1@^1.1.3:
```
[the remark this answers is not in the document]
Comment (najla-quant) [reply yarn.lock]: kept

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts (hunk 3 above)
Comment (najla-quant) [reply packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: parseInt

File: packages/overledger-dlt-bitcoin/src/Bitcoin.ts (hunk 4 above)
Comment (lukerQuant) [reply packages/overledger-dlt-bitcoin/src/Bitcoin.ts]: It is incorporated in the accounts object (as a subType)

**Pull request merged**